core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -122,6 +122,30 @@ public class TaskMemoryManager {
    */
   private volatile long acquiredButNotUsed = 0L;
 
+  /**
+   * Current off heap memory usage by this task.
+   */
+  private long currentOffHeapMemory = 0L;
+
+  private final Object offHeapMemoryLock = new Object();
+
+  /*
+   * Current on heap memory usage by this task.
+   */
+  private long currentOnHeapMemory = 0L;
```

Comment:
Use `memoryManager.<foo>MemoryUsed` can be used instead ?

Reply:
Do you mean `memoryManager.getOffHeapExecutionMemoryUsageForTask` ? That was the [original implementation](https://github.com/apache/spark/pull/47192/commits/8cec467fb563b3e8cef5d7281665f5db19a3786d#diff-dd7c6261f4882a0c8d5369ac960267154f7f4dde991f491311f22e249586e35bL218), but that require synchronization on `MemoryManager` granularity, so we decided to maintain this on `TaskMemoryManager` level (see this [discussion thread](https://github.com/apache/spark/pull/47192#discussion_r1678527386))

Reply:
I would prefer removing duplication given this is already maintained in MemoryManager.
The change will then become something like:

In acquireExecutionMemory:
```
  updatePeakStats(mode);
  return got;
```

```

private void updatePeakStats(MemoryMode mode) {
  // this precondition should hold - though we dont need to necessarily assert it.
  assert(Thread.holdsLock(this));
  if (mode == MemoryMode.OFF_HEAP) {
    peakOffHeapMemory = Math.max(peakOffHeapMemory, memoryManager.offHeapExecutionMemoryUsed);
  } else {
    peakOnHeapMemory = Math.max(peakOnHeapMemory, memoryManager.onHeapExecutionMemoryUsed);
  }
}
```

In releaseExecutionMemory:
```
synchronized(this) {
  memoryManager.releaseExecutionMemory(size, taskAttemptId, consumer.getMode());
  updatePeakStats(consumer.getMode());
}
```

Thoughts ?
+CC @JoshRosen, @Ngone51.